### PR TITLE
Fix 833 - if_else interaction between SIMD and scalar product_type

### PR DESCRIPTION
**Pull request by jfalcou**

### CODE REVIEW

File: include/eve/detail/meta.hpp [hunk 1]
```diff
@@ -367,6 +367,15 @@ namespace eve::detail
...
   {};
 
+  // Get the first element of something
+  template<typename T> struct first_of { using type = T; };
+
+  template<typename T>
+  requires(kumi::product_type<T>) struct first_of<T> : kumi::element<0,T>
+  {};
+
+  template<typename T> using first_of_t = typename first_of<T>::type;
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/meta.hpp:377]:
```
std::tuple_element<0, T>
```
?

File: include/eve/module/real/core/function/regular/generic/if_else.hpp [hunk 2]
```diff
@@ -53,11 +55,26 @@ namespace eve::detail
...
+            if constexpr(has_emulated_abi_v<T>  ) return map(if_else, cond, r_t(t), r_t(f));
+      else  if constexpr(has_aggregated_abi_v<T>) return aggregate(if_else, cond, r_t(t), r_t(f));
       else  if constexpr(std::same_as<logical<e_t>,element_type_t<T>>)
       {
-        if constexpr( std::same_as<U,V> ) return  bit_select(cond.mask(),r_t(t), r_t(f));
+        if constexpr( std::same_as<U,V> )
+        {
+          if constexpr(kumi::product_type<U> || kumi::product_type<V>)
+          {
+            if constexpr( detail::tag_dispatchable<tag::if_else_,T,U,V> )
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/core/function/regular/generic/if_else.hpp:66]: why? What could possible be a reason to customise `if_else`?

File: include/eve/module/real/core/function/regular/generic/if_else.hpp [hunk 3]
```diff
@@ -53,11 +55,26 @@ namespace eve::detail
...
+            {
+              return tagged_dispatch(tag::if_else_{}, cond, t, f);
+            }
+            else
+            {
+              using   w_t = std::conditional_t< simd_value<U>, U, V>;
+              return  w_t{ kumi::map([&](auto v, auto f) { return if_else(cond,v,f); }, t, f) };
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/core/function/regular/generic/if_else.hpp:73]: `tuple_select`? You have it for some reason?

File: include/eve/module/real/core/function/regular/generic/if_else.hpp [hunk 4]
```diff
@@ -90,9 +104,7 @@ namespace eve::detail
   //------------------------------------------------------------------------------------------------
   // Optimizes if_else(c,t,constant)
   template<value T, value U, typename Constant>
-  requires( ( kumi::product_type<U> && std::is_invocable_v<Constant, as<kumi::element_t<0,U>>> )
-          || std::is_invocable_v<Constant, as<U>>
-          )
+  requires( !kumi::product_type<Constant> && std::invocable<Constant, as<first_of_t<U>>> )
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/core/function/regular/generic/if_else.hpp:107]: Why only first_of?

File: include/eve/module/real/core/function/regular/generic/if_else.hpp (hunk 2 above)
Comment (jfalcou) [reply include/eve/module/real/core/function/regular/generic/if_else.hpp:66]: you have a custom data type from which some member has no business being selected via a mask cause it is only some internal buffers. It cost nothing to do this check anyway so we can have it.

File: include/eve/module/real/core/function/regular/generic/if_else.hpp (hunk 3 above)
Comment (jfalcou) [reply include/eve/module/real/core/function/regular/generic/if_else.hpp:73]: I forgot to remove tuple_select as in the end, it only get used once.

File: include/eve/module/real/core/function/regular/generic/if_else.hpp (hunk 4 above)
Comment (jfalcou) [reply include/eve/module/real/core/function/regular/generic/if_else.hpp:107]: I have no easy way here to iterate over all the member at this place. I can add a traits doing this if needed.

File: include/eve/detail/meta.hpp (hunk 1 above)
Comment (jfalcou) [reply include/eve/detail/meta.hpp:377]: I like the fact we keep using kumi API with kumi types.

File: include/eve/module/real/core/function/regular/generic/if_else.hpp (hunk 4 above)
Comment (DenisYaroshevskiy) [reply include/eve/module/real/core/function/regular/generic/if_else.hpp:107]: Ah. Well whatever.

`kumi::element_t<0, T>`?

File: include/eve/detail/meta.hpp [hunk 5]
```diff
@@ -367,6 +368,32 @@ namespace eve::detail
...
+  template<typename Constant, typename Type>
+  struct is_generator : std::is_invocable<Constant,as<Type>>
+  {};
+
+  template<typename Constant, typename Type, typename Index>
+  struct  is_generator_impl;
+
+  template<typename Constant, typename Type, std::size_t... I>
+  struct  is_generator_impl<Constant,Type,std::index_sequence<I...>>
+        : std::conjunction<is_generator<Constant, std::tuple_element_t<I,Type>>...>
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/meta.hpp:381]: just use a fold expression, it's faster to compile

File: include/eve/detail/meta.hpp [hunk 6]
```diff
@@ -367,6 +368,32 @@ namespace eve::detail
...
+  struct  is_generator<Constant,Type>
+        : is_generator_impl<Constant,Type,std::make_index_sequence<std::tuple_size<Type>::value>>
+  {};
+
+
+  template<typename Constant, typename Type>
+  using is_generator_t = typename is_generator<Constant,Type>::type;
+
+  template<typename Constant, typename Type>
+  inline constexpr bool is_generator_v = is_generator<Constant,Type>::value;
```
Comment (DenisYaroshevskiy) [inline include/eve/detail/meta.hpp:395]: This should be a concept

File: include/eve/module/real/core/function/regular/generic/if_else.hpp [hunk 7]
```diff
@@ -53,8 +56,9 @@ namespace eve::detail
       using e_t = element_type_t<common_compatible_t<U, V>>;
       using r_t = as_wide_t<e_t, cardinal_t<T>>;
 
-            if constexpr(has_emulated_abi_v<T>)    return map(if_else, cond, r_t(t), r_t(f));
-      else  if constexpr(has_aggregated_abi_v<T>)  return aggregate(if_else, cond, r_t(t), r_t(f));
+            if constexpr(has_emulated_abi_v<T>  ) return map(if_else, cond, r_t(t), r_t(f));
+      else  if constexpr(has_aggregated_abi_v<T>) return aggregate(if_else, cond, r_t(t), r_t(f));
+      else  if constexpr(kumi::product_type<U> || kumi::product_type<V>) return tuple_select(cond,t,f);
```
Comment (DenisYaroshevskiy) [inline include/eve/module/real/core/function/regular/generic/if_else.hpp:61]: You have an `||` here but `tuple_select` needs 2 product types.
Missing tests?

File: include/eve/detail/meta.hpp (hunk 6 above)
Comment (jfalcou) [reply include/eve/detail/meta.hpp:395]: Done

File: include/eve/detail/meta.hpp (hunk 5 above)
Comment (jfalcou) [reply include/eve/detail/meta.hpp:381]: Done

File: include/eve/conditional.hpp [hunk 8]
```diff
@@ -184,7 +182,7 @@ namespace eve
       }
       else
       {
-        using i_t = as_arithmetic_t<as_integer_t<T>>;
+        using i_t = as_integer_t<typename type::mask_type>;
```
Comment (DenisYaroshevskiy) [inline include/eve/conditional.hpp:185]: Can mask type not be integer?

File: include/eve/traits/as_logical.hpp [hunk 9]
```diff
@@ -29,6 +30,12 @@ namespace eve
     using type = logical<T>;
   };
 
+
+  template<typename T>
+  requires( kumi::product_type<T> )
+  struct as_logical<T> : as_logical< kumi::element_t<0,T> >
+  {};
```
Comment (DenisYaroshevskiy) [inline include/eve/traits/as_logical.hpp:37]: This is arguable. But we can change later, w/e

**Review by DenisYaroshevskiy — APPROVED** [review]

**Pull request merged**